Implicit `DEFAULT_LIMIT` should be 8 bytes (64-bit signed integer)

kamipo opened the pull request:
SQLite integer type is stored in 8 bytes. Assuming as 4 bytes causes
that valid in range value cannot be stored and existing value cannot be
found.

https://www.sqlite.org/datatype3.html

Fixes #22594.

kamipo committed:
Implicit `DEFAULT_LIMIT` should be 8 bytes (64-bit signed integer)

SQLite integer is stored in 1, 2, 3, 4, 6, or 8 bytes depending on the
magnitude of the value. Assuming as 4 bytes causes that valid in range
value cannot be stored and existing value cannot be found.

https://www.sqlite.org/datatype3.html

Fixes #22594.

kamipo commented:
@rafaelfranca @matthewd Any chance this fix could be merged in 5.1?

matthewd commented:
I'm not sure I agree 8 is a better default for Integer. In most places that integer size distinctions are made, I think it normally still means a 32-bit value.

This seems to be something the SQLite adapter should be overriding because *it* is unusual by treating everything as a bigint.

kamipo commented:
I've created #28379 as an alternative of #28007.

Originally this PR is a part of #26302 to eliminate `RangeError` in `ActiveModel`.

Assuming `DEFAULT_LIMIT` and raising `RangeError` were introduced since #17459 in order to ease to handle out of range numbers for `find` and `find_by` in postgresql adapter with prepared statements.

But the raising error has broken some features (https://github.com/rails/rails/pull/25230/commits/b8928d7c8b602e7b1e8a52e0e666c3b41fb62251). And it also bothers many users (https://github.com/Shopify/identity_cache/pull/299, #18134, #18787, #19717, #20140, #20428, #21309, #22594, #24238, and more). I've sometimes addressed this problem (#17793, #25271, #25280).

As a better solution, I implemented that fallback to unprepared statement instead of raising `RangeError` (https://github.com/rails/rails/pull/25230/commits/5021ce4609b7e0d4e8db73452dd94346bd3ffc77). This implementation fixes these use case (https://github.com/rails/rails/pull/25230/commits/b8928d7c8b602e7b1e8a52e0e666c3b41fb62251).

Currently `ActiveRecord::Type` was moved to `ActiveModel` (9cc8c6f3730). And moved back database-specific types (#26696). I think that raising `RangeError` is a database-specific feature. `ActiveModel` should not raise `RangeError` against valid integer value in Ruby.

```
Loading development environment (Rails 5.1.0.beta1)
irb(main):001:0> type = ActiveModel::Type.lookup(:integer)
=> #<ActiveModel::Type::Integer:0x007f850967e860 @precision=nil, @scale=nil, @limit=nil, @range=-2147483648...2147483648>
irb(main):002:0> 1<<128
=> 340282366920938463463374607431768211456
irb(main):003:0> (1<<128).class
=> Integer
irb(main):004:0> type.serialize(1<<128)
ActiveModel::RangeError: 340282366920938463463374607431768211456 is out of range for ActiveModel::Type::Integer with limit 4 bytes
	from (irb):4
```

Farranco commented:
I am also running into this issue, but I am using Heoku, so can't change the source of Padrino running. How do I override the default in my app source? I have a bounty on stackoverflow if anyone can help: https://stackoverflow.com/questions/43431244/type-casting-a-column-to-bigint-in-padrino-active-record

kamipo commented:
Closing this in favor of #28379.
